### DISCUSSION TITLE
Question - About presence feauture

### DISCUSSION THREAD

**Discussion by mhtb32** [eleurent/highway-env#172] [General]
I looked up the code but couldn't figure out what is the presence feature:

- When does its value change? since in `Vehicle.to_dict()` method always `1` is returned.
- What is the use of it?

Regards.

**Comment by eleurent** [comment]
The `presence` feature is mainly used in the [KinematicsObservation](https://github.com/eleurent/highway-env/blob/master/highway_env/envs/common/observation.py#L96).
It is a byproduct of the constraint that np.array observations need to have fixed size: N rows are allocated to describe vehicles, but there may be n<N vehicles in the scene (or nearby). A presence flag at 0 is used to indicate that the row is only a placeholder, and that the position/speed information are meaningless.

**Comment by mhtb32** [comment]
So is it up to the RL agent to learn from its experience that those observations with `presence == 1` are meaningless and do not contribute to achieving reward? Or these observations must be preprocessed before feeding to the agent?

**Comment by eleurent** [comment]
Both are possible, however function approximation often also expects fixed-size inputs (e.g. feedfoward neural networks), which kind of prevents this sort of preprocessing (i guess you meant filtering out absent vehicles).

**Comment by mhtb32** [comment]
Yes, I meant that.
